spec/v3/spec.md

```diff
@@ -1,93 +1,372 @@
-| 16 | length of root directory | 8 |
-| 24 | offset of JSON metadata, possibly compressed by `InternalCompression` | 8 |
-| 32 | length of JSON metadata | 8 |
-| 40 | offset of leaf directories | 8 |
-| 48 | length of leaf directories | 8 |
-| 56 | offset of tile data | 8 |
-| 64 | length of tile data | 8 |
-| 72 | # of addressed tiles, 0 if unknown | 8 |
-| 80 | # of tile entries, 0 if unknown | 8 |
-| 88 | # of tile contents, 0 if unknown | 8 |
-| 96 | boolean clustered flag, `0x1` if true | 1 |
-| 97 | `InternalCompression` enum (0 = Unknown, 1 = None, 2 = Gzip, 3 = Brotli, 4 = Zstd) | 1 |
-| 98 | `TileCompression` enum | 1 |
-| 99 | tile type enum (0 = Unknown/Other, 1 = MVT (PBF Vector Tile), 2 = PNG, 3 = JPEG, 4 = WEBP | 1 |
-| 100 | min zoom | 1 |
-| 101 | max zoom | 1 |
-| 102 | min longitude (signed 32-bit integer: longitude * 10,000,000) | 4 |
-| 106 | min latitude | 4 |
-| 110 | max longitude | 4 |
-| 114 | max latitude | 4 |
-| 118 | center zoom | 1 |
-| 119 | center longitude | 4 |
-| 123 | center latitude | 4 |
-
-### Notes
-
-* **# of addressed tiles**: the total number of tiles before run-length encoding, i.e. `Sum(RunLength)` over all entries.
-* **# of tile entries**: the total number of entries across all directories where `RunLength > 0`.
-* **# # of tile contents**: the number of referenced blobs in the tile section, or the unique # of offsets. If the archive is completely deduplicated, this is equal to the # of unique tile contents. If there is no deduplication, this is equal to the number of tile entries above.
-* **boolean clustered flag**: if true, blobs in the data section are ordered by Hilbert `TileId`. When writing with deduplication, this means that offsets are either contiguous with the previous offset+length, or refer to a lesser offset.
-* **compression enum**: Mandatory, tells the client how to decompress contents as well as provide correct `Content-Encoding` headers to browsers.
-* **tile type**: A hint as to the tile contents. Clients and proxies may use this to:
- 	* Automatically determine a visualization method
-	* provide a conventional MIME type `Content-Type` HTTP header
-	* Enforce a canonical extension e.g. `.mvt`, `png`, `jpeg`, `.webp` to prevent duplication in caches
-
-### Organization
-
-In most cases, the archive should be in the order `Header`, Root Directory, JSON Metadata, Leaf Directories, Tile Data. It is possible to relocate sections other than `Header` arbitrarily, but no current writers/readers take advantage of this. A future design may allow for reverse-ordered archives to enable single-pass writing.
+# PMTiles Version 3 Specification
+
+## 1 Introduction
+
+PMTiles is a single-file archive of square tiles.
+
+## 2 Overview
+
+A archive consist of five main sections:
+
+1. A fixed-size 127-byte header (described in [chapter 3](#3-header))
+1. A root directory (described in [chapter 4](#4-directories))
+1. Optional JSON metadata (described in [chapter 5](#5-json-metadata))
```

Comment:
@bdon in https://github.com/protomaps/PMTiles/issues/107#issuecomment-1535056512:

> I think we should clarify the spec to make JSON required: the null case is just the empty JSON object. I don't see any downside to doing this and it simplifies the client.

Reply:
Fine by me. I remember running into this when trying out my own encoder.
Can this be any valid JSON value or does this have to be a object? 42, null or an array would also be valid JSON 🤔I personally would welcome that. You could make a good case about at least supporting top level arrays and if you support arrays and objects you might as well allow any valid JSON 😅

Reply:
I backported the changes from 3.1 in 60cb7259e, which requires it to be an object.